File: packages/api/src/schemas/decisions.py
"""
Decision and risk assessment Pydantic schemas.
"""

from datetime import datetime

from pydantic import BaseModel


class RiskDimensionScoreResponse(BaseModel):
    """Individual risk dimension score."""

    name: str
    score: float
    weight: float
    weighted_score: float
    agent: str
    positive_factors: list[str] = []
    risk_factors: list[str] = []
    mitigating_factors: list[str] = []
    explanation: str | None = None


class RiskCondition(BaseModel):
    """A condition for conditional approval."""

    condition: str
    status: str = "pending"


class RiskAssessmentResponse(BaseModel):
    """Full risk assessment response (servicer view)."""

    id: str
    application_id: str
    status: str
    overall_score: float | None = None
    risk_band: str | None = None
    confidence: float | None = None
    recommendation: str | None = None
    summary: str | None = None
    dimensions: list[RiskDimensionScoreResponse] = []
    conditions: list[RiskCondition] = []
    processing_metadata: dict | None = None


class RiskAssessmentApplicantView(BaseModel):
    """Simplified risk assessment response (applicant view)."""

    id: str
    status: str
    overall_score: float | None = None
    risk_band: str | None = None
    recommendation: str | None = None
    summary: str | None = None
    positive_highlights: list[str] = []
    areas_of_concern: list[str] = []


class DecisionCreate(BaseModel):
    """Request to create a loan decision."""

    decision: str  # approved, denied, conditionally_approved
    conditions: list[str] = []
    adverse_action_reasons: list[str] = []
    interest_rate: float | None = None
    approved_loan_amount: float | None = None
    approved_term_years: int | None = None
    notes: str | None = None
    override_ai_recommendation: bool = False
    override_justification: str | None = None


class DecisionResponse(BaseModel):
    """Decision response."""

    id: str
    application_id: str
    decision: str
    ai_recommendation: str | None = None
    servicer_agreed_with_ai: bool | None = None
    override_justification: str | None = None
    conditions: list[str] = []
    adverse_action_reasons: list[str] = []
    interest_rate: float | None = None
    approved_loan_amount: float | None = None
    approved_term_years: int | None = None
    monthly_payment: float | None = None
    notes: str | None = None
    explanation: str | None = None
    decided_by_name: str | None = None
    decided_by_role: str | None = None
    decided_at: datetime | None = None


class InfoRequestItem(BaseModel):
    """Single item in an information request."""

    type: str  # document, clarification
    document_type: str | None = None
    description: str | None = None
    question: str | None = None


class InfoRequestCreate(BaseModel):
    """Request additional information from applicant."""

    requested_items: list[InfoRequestItem]
    due_date: str | None = None


class InfoRequestResponse(BaseModel):
    """Response after creating an info request."""

    application_id: str
    status: str
    requested_items: list[InfoRequestItem]
    due_date: str | None = None
    message: str

File: packages/ui/src/schemas/decisions.ts
import { z } from 'zod';

export const RiskDimensionScoreSchema = z.object({
  name: z.string(),
  score: z.number(),
  weight: z.number(),
  weighted_score: z.number(),
  agent: z.string(),
  positive_factors: z.array(z.string()),
  risk_factors: z.array(z.string()),
  mitigating_factors: z.array(z.string()),
  explanation: z.string().nullable(),
});

export const RiskConditionSchema = z.object({
  condition: z.string(),
  status: z.string().default('pending'),
});

export const RiskAssessmentSchema = z.object({
  id: z.string(),
  application_id: z.string(),
  status: z.string(),
  overall_score: z.number().nullable(),
  risk_band: z.string().nullable(),
  confidence: z.number().nullable(),
  recommendation: z.string().nullable(),
  summary: z.string().nullable(),
  dimensions: z.array(RiskDimensionScoreSchema).default([]),
  conditions: z.array(RiskConditionSchema).default([]),
  processing_metadata: z.record(z.unknown()).nullable(),
});

export const DecisionSchema = z.object({
  id: z.string(),
  application_id: z.string(),
  decision: z.string(),
  ai_recommendation: z.string().nullable(),
  servicer_agreed_with_ai: z.boolean().nullable(),
  override_justification: z.string().nullable(),
  conditions: z.array(z.string()),
  adverse_action_reasons: z.array(z.string()),
  interest_rate: z.number().nullable(),
  approved_loan_amount: z.number().nullable(),
  approved_term_years: z.number().nullable(),
  monthly_payment: z.number().nullable(),
  notes: z.string().nullable(),
  explanation: z.string().nullable(),
  decided_by_name: z.string().nullable(),
  decided_by_role: z.string().nullable(),
  decided_at: z.string().nullable(),
});

export const DashboardStatsSchema = z.object({
  pending_review: z.number(),
  in_progress: z.number(),
  decided_today: z.number(),
  average_processing_time_hours: z.number().nullable(),
  approval_rate: z.number().nullable(),
  risk_distribution: z.object({
    low: z.number(),
    medium: z.number(),
    high: z.number(),
    very_high: z.number(),
  }),
  recent_applications: z.array(
    z.object({
      id: z.string(),
      application_number: z.string(),
      status: z.string(),
      applicant_name: z.string().nullable(),
      loan_type: z.string().nullable(),
      loan_amount: z.number().nullable(),
      risk_score: z.number().nullable(),
      risk_band: z.string().nullable(),
      submitted_at: z.string().nullable(),
      created_at: z.string().nullable(),
      assigned_servicer: z.string().nullable(),
    }),
  ),
});

export type RiskDimensionScore = z.infer<typeof RiskDimensionScoreSchema>;
export type RiskCondition = z.infer<typeof RiskConditionSchema>;
export type RiskAssessment = z.infer<typeof RiskAssessmentSchema>;
export type Decision = z.infer<typeof DecisionSchema>;
export type DashboardStats = z.infer<typeof DashboardStatsSchema>;

export interface DecisionCreateInput {
  decision: 'approved' | 'denied' | 'conditionally_approved';
  conditions?: string[];
  adverse_action_reasons?: string[];
  interest_rate?: number;
  approved_loan_amount?: number;
  approved_term_years?: number;
  notes?: string;
  override_ai_recommendation?: boolean;
  override_justification?: string;
}

export interface InfoRequestItem {
  type: 'document' | 'clarification';
  document_type?: string;
  description?: string;
  question?: string;
}

export interface InfoRequestInput {
  requested_items: InfoRequestItem[];
  due_date?: string;
}
